aten/src/ATen/native/quantized/cpu/qnnpack_relu.cpp

```diff
@@ -66,8 +67,10 @@ class QNNPACKRelu final : public torch::OperatorKernel {
         setupStatus == qnnp_status_success,
         "failed to setup QNNPACK Relu operator");
 
+    pthreadpool_t threadpool = at::mobile_threadpool();
+
     const qnnp_status runStatus =
-        qnnp_run_operator(qnnpack_operator, nullptr /* thread pool */);
+        qnnp_run_operator(qnnpack_operator, threadpool);
```

Comment:
I believe a better approach would be to not pass thread pool as an explicit argument but have an operator itself call `at::parallel_for` and analogues, @supriyar have you checked what primitives are actually used in qnnpack implementations? We can add extra ones in ATen/Parallel.h and I can add a backend of ATen/Parallel.h specifically for mobile

Reply:
I totally agree that this is not compatible with the idea of aten::parallel - we should add more context to the PR description to give more context / future work plan.

- How things work for caffe2:
Caffe2 Ops -> NNPACK/QNNPACK -> pthreadpool_compute_1/2/3/4d_tiled -> pthreadpool_compute_1d (caffe2 shim) -> caffe2::ThreadPool

- Before this PR:
Pytorch Ops -> NNPACK/QNNPACK -> pthreadpool_compute_1/2/3/4d_tiled -> pthreadpool_compute_1d (third_party implementation without mobile optimization)

So our work plan is:
1. Take a baby step, make sure libtorch mobile links against caffe2/util/pthreadpool with caffe2::ThreadPool instead of the 3rd party one.

2. Implement a mobile version of "at::parallel_for" using caffe2::ThreadPool so all ATen/TH multithreading usage is mobile friendly;

3. Refactor QNNPACK or pthreadpool to explicitly using "at::parallel_for" primitive to replace pthreadpool_compute_1d for Pytorch;

When we get to step 3, we will delete the mobile_threadpool() API. This is just a temporary step to unblock us from running multi-thread QNNPack perf tests.

I feel the main reason to not pass thread-pool explicitly here is to hide threadpool as implementation detail. Whether to create threadpool upfront or lazily is somewhat orthogonal (see my comment about abstract resource management API above).

We could rename it to something like "_mobile_threadpool" or move it to a different util class without "ParallelXXX" prefix - but it's really just a temporary thing :)

Reply:
This explanation and plan makes sense, thanks! Definitely please copy paste in the PR description

Reply:
Updated the PR description.